omf-python/src/colormap.rs

```diff
@@ -0,0 +1,212 @@
+use crate::array::{PyBoundaryArray, PyGradientArray};
+use omf::{NumberColormap, NumberRange};
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+use pyo3_stub_gen::derive::*;
+
+#[gen_stub_pyclass]
+#[pyclass(name = "NumberRangeFloat")]
+/// Specifies the minimum and maximum values of a number colormap.
+///
+/// Values outside this range will use the color at the ends of the gradient.
+/// The variant used should match the type of the number array.
+pub struct PyNumberRangeFloat(pub NumberRange);
+
+#[gen_stub_pymethods]
+#[pymethods]
+impl PyNumberRangeFloat {
+    #[getter]
+    fn min(&self) -> PyResult<f64> {
+        match &self.0 {
+            NumberRange::Float { min, .. } => Ok(*min),
+            _ => Err(PyValueError::new_err(
+                "NumberRange variant is not supported",
+            )),
+        }
+    }
+    #[getter]
+    fn max(&self) -> PyResult<f64> {
+        match &self.0 {
+            NumberRange::Float { max, .. } => Ok(*max),
+            _ => Err(PyValueError::new_err(
+                "NumberRange variant is not supported",
+            )),
+        }
+    }
+}
+
+#[gen_stub_pyclass]
+#[pyclass(name = "NumberRangeInteger")]
+/// Specifies the minimum and maximum values of a number colormap.
+///
+/// Values outside this range will use the color at the ends of the gradient.
+/// The variant used should match the type of the number array.
+pub struct PyNumberRangeInteger(pub NumberRange);
+
+#[gen_stub_pymethods]
+#[pymethods]
+impl PyNumberRangeInteger {
+    #[getter]
+    fn min(&self) -> PyResult<i64> {
+        match &self.0 {
+            NumberRange::Integer { min, .. } => Ok(*min),
+            _ => Err(PyValueError::new_err(
+                "NumberRange variant is not supported",
+            )),
+        }
+    }
+    #[getter]
+    fn max(&self) -> PyResult<i64> {
+        match &self.0 {
+            NumberRange::Integer { max, .. } => Ok(*max),
+            _ => Err(PyValueError::new_err(
+                "NumberRange variant is not supported",
+            )),
+        }
+    }
+}
+
+#[gen_stub_pyclass]
+#[pyclass(name = "NumberRangeDate")]
+/// Specifies the minimum and maximum values of a number colormap.
+///
+/// Values outside this range will use the color at the ends of the gradient.
+/// The variant used should match the type of the number array.
+pub struct PyNumberRangeDate(pub NumberRange);
+
+#[gen_stub_pymethods]
+#[pymethods]
+impl PyNumberRangeDate {
+    #[getter]
+    fn min(&self, py: Python<'_>) -> PyResult<PyObject> {
+        match &self.0 {
+            NumberRange::Date { min, .. } => Ok((*min).into_py(py)),
```

Comment:
Do you need to use into_py() and PyObject here as you only have on return type? Can you just return *min ?

Reply:
I think because `min` is a `chrono::datetime::DateTime<chrono::offset::utc::Utc>` we need `into_py()` to cast it to a Python datetime object using the chrono extension https://docs.rs/pyo3/latest/pyo3/chrono/index.html#

At least that's the intent anyway. 

I tried removing the into_py and the compiler complains about mismatched types, so I tried changing the return type of the function `PyResult<DateTime<Utc>>` but then I get `the trait bound `chrono::DateTime<Utc>: PyStubType` is not satisfied`.

I think you're probably right, but my current knowledge isn't sufficient for finding a way to do it.

Also, yeah I thought we had come across this dereferencing stuff before- good catch. I also tried applying that pattern to `PyNumberColormapContinuous` but then had to add back the `.clone()`s, so not sure which way to go on that one